### DISCUSSION TITLE
Identity Backup and Restore

### DISCUSSION THREAD

**Discussion by iPaulPro** [deso-protocol/dips#104] [Medium CIPs]
---

cip: <to be assigned>
title: Identity Backup and Restore
author: Paul Burke (@ipaulpro)
discussions-url: [#104](https://github.com/bitclout/cips/discussions/104)
created: 2021-08-01

---

## One-Line Description

A method is proposed for backup and recovery of a user's private key, which allows an authenticated user to generate a "backup file" that can subsequently be used to "log in" with Identity.

## Explanation and Motivation
The motivation for this proposal stems from the desire to provide alternative methods of authentication beyond entering a mnemonic phrase. It allows authenticated users to recover their account in the event of a _forgotten_ mnemonic.

## Specification
The `hdkey` library provides the functions `toJSON()` and `fromJSON(obj)` as a means of creating a keychain from a master private key. This proposals suggests adding an affordance to Identity, for authenticated users, that provides the ability for a user to download the data returned from `HDKey.toJSON()` as a file. Consequently, an alternative means for "login" should be added to allow for the submission of the aforementioned file.

![keystore](https://user-images.githubusercontent.com/354227/127790536-4b29bd8e-5a10-4fcb-ac95-bf4de82dd69e.png)

### Data Storage Change
None

## Backwards Compatibility
No existing Identity authentication functions are altered with this addition.

## Security Considerations
Private key backup files are generally considered insecure. To improve security, a password could be used to encrypt the exported JSON file, creating a Keystore file.

## Acknowledgements
Thanks to @carsenk, @BogdanDidenko, and @maebeam

**Comment by omicreativedev** [comment]
Popping in to say I think this is a great idea.

**Comment by tijno** [comment]
Loving this too @iPaulPro - but it would be even more amazing if the file could be encrypted with a password? That way people dont even need to worry about storing it securely.

Eg the issue with Arweave for me is that you download the key as plaintext json which I then need to encrypt myself, and make sure I delete the original and clear trash. Even then it could probably still be recovered from my machine.

**Reply by iPaulPro** [reply]
> Loving this too @iPaulPro - but it would be even more amazing if the file could be encrypted with a password? That way people dont even need to worry about storing it securely.

Indeed, see the "Security Considerations" section. I think this is an important option, and perhaps a crucial one. Maybe I should add more details to the specification on it.

**Reply by tijno** [reply]
note to self - read better

**Comment by sungkhum** [comment]
Great idea
